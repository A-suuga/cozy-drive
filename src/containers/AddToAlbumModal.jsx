import styles from '../styles/addToAlbum'

import React from 'react'
import { connect } from 'react-redux'
import { translate } from '../lib/I18n'
import Modal from 'cozy-ui/react/Modal'
import classNames from 'classnames'

import Alerter from '../components/Alerter'
import CreateAlbumForm from '../components/CreateAlbumForm'

import { cancelAddToAlbum, createAlbum, addToAlbum } from '../actions/albums'

const AddToAlbumModal = ({t, mangoIndex, photos,
  onDismiss, onSubmitNewAlbum }) => {
  return (
    <Modal
      title={t('Albums.add_photos.title')}
      secondaryAction={() => onDismiss()}
      >
      <div className={classNames(styles['coz-modal-section'], styles['coz-create-album'])}>
        <CreateAlbumForm
          onSubmitNewAlbum={(name) => onSubmitNewAlbum(name, mangoIndex, photos)}
          />
      </div>
    </Modal>
  )
}

const mapStateToProps = (state, ownProps) => {
  return {
    photos: state.ui.selected,
    mangoIndex: state.mango.albumsIndexByName
  }
}

const handleActionError = name => {
  return error => {
    const isUnexpectedError = !!error.message
    Alerter.error(isUnexpectedError ? 'Albums.add_photos.error.generic' : error, {name: name})
    return Promise.reject(error)
  }
}

const mapDispatchToProps = (dispatch, ownProps) => ({
  onDismiss: () => {
    dispatch(cancelAddToAlbum())
  },
  // Removes photos parameter when we will be able to pick an album instead
  // of adding photo to the created one by default.
  onSubmitNewAlbum: (name, mangoIndex, photos) => {
    return dispatch(createAlbum(name, mangoIndex, photos))
      .then(
        album => {
          dispatch(addToAlbum(photos, album))
            .then(() => Alerter.success('Albums.add_photos.success', {name: album.name, smart_count: photos.length}))
            .catch(handleActionError)
        },
        handleActionError(name)
      )
  }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(translate()(AddToAlbumModal))
