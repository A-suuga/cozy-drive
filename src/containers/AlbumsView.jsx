import React, { Component } from 'react'
import { connect } from 'react-redux'
import styles from '../styles/layout'

import { AlbumsToolbar, fetchAlbums, getAlbumsList } from '../ducks/albums'
import { filterSharedDocuments } from '../ducks/sharing'

import AlbumsList from '../components/AlbumsList'
import Loading from '../components/Loading'
import ErrorComponent from '../components/ErrorComponent'
import Topbar from '../components/Topbar'

const Content = ({ list, shared }) => {
  const { fetchStatus, entries } = list
  switch (fetchStatus) {
    case 'pending':
    case 'loading':
      return <Loading loadingType='albums_fetching' />
    case 'failed':
      return <ErrorComponent errorType='albums' />
    default:
      return <AlbumsList albums={entries} shared={shared} />
  }
}

export class AlbumsView extends Component {
  constructor (props) {
    super(props)
    this.state = {
      shared: []
    }
  }

  componentWillMount () {
    this.props.fetchAlbums()
  }

  componentWillReceiveProps (newProps) {
    // TODO: this is not the cleanest way of doing this...
    if (newProps.albums && newProps.albums.entries !== 0) {
      filterSharedDocuments(newProps.albums.entries.map(a => a.id), 'io.cozy.photos.albums')
        .then(ids => this.setState({ shared: ids }))
    }
  }

  render () {
    if (this.props.children) return this.props.children
    if (!this.props.albums) {
      return null
    }
    return (
      <div className={styles['pho-content-wrapper']}>
        <Topbar viewName='albums'>
          <AlbumsToolbar />
        </Topbar>
        <Content list={this.props.albums} shared={this.state.shared} />
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => ({
  albums: getAlbumsList(state)
})

export const mapDispatchToProps = (dispatch, ownProps) => ({
  fetchAlbums: () => dispatch(fetchAlbums())
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AlbumsView)
