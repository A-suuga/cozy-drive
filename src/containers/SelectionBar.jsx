import styles from '../styles/selectionbar'

import React from 'react'
import { connect } from 'react-redux'
import classNames from 'classnames'
import { translate } from '../lib/I18n'

import { hideSelectionBar } from '../actions/selection'
import { addToAlbum } from '../actions/albums'

const SelectionBar = ({ t, selected, selectedCount, onHide, onAddToAlbum }) => (
  <div
    className={classNames(styles['coz-selectionbar'], {
      [styles['coz-selectionbar--active']]: selectedCount !== 0
    })}
    role='toolbar'
  >
    <span className={styles['coz-selectionbar-count']}>
      {t('SelectionBar.selected_count', { smart_count: selectedCount })}
    </span>
    <span className={styles['coz-selectionbar-separator']} />
    <button
      disabled={selectedCount === 0}
      className={styles['pho-action-album-add']}
      onClick={onAddToAlbum(selected)}
    >
      {t('SelectionBar.add_to_album')}
    </button>
    <button className={styles['coz-action-close']} onClick={onHide}>
      {t('SelectionBar.close')}
    </button>
  </div>
)

const mapStateToProps = (state, ownProps) => ({
  selected: state.ui.selected,
  selectedCount: state.ui.selected.length
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  onHide: () => {
    dispatch(hideSelectionBar())
  },
  onAddToAlbum: (selected) => {
    return (event) => {
      event.preventDefault()
      dispatch(addToAlbum(selected))
    }
  }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(translate()(SelectionBar))
