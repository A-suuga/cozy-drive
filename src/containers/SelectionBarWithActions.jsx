import React from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'
import { translate } from 'cozy-ui/react/I18n'
import { openAddToAlbum, removeFromAlbum } from '../ducks/albums'
import { SelectionBar } from '../ducks/selection'
import Alerter from '../components/Alerter'
import confirm from '../lib/confirm'
import { DeleteConfirm, deletePhotos } from '../ducks/timeline'

const mapDispatchToProps = (dispatch, ownProps) => ({
  actions: {
    'album-add': {
      action: selected => dispatch(openAddToAlbum(selected))
    },
    'delete': {
      action: selected => confirm(<DeleteConfirm t={ownProps.t} count={selected.length} related={false} />,
        () => dispatch(deletePhotos(selected))
      ),
      displayCondition: () => ownProps.router.location.pathname.startsWith('/photos')
    },
    'album-remove': {
      action: selected => dispatch(removeFromAlbum(ownProps.album, selected))
        .then(() => Alerter.success('Albums.remove_photos.success', { album_name: ownProps.album.name }))
        .catch(() => Alerter.error('Albums.remove_photos.error.generic')),
      displayCondition: () => ownProps.album !== undefined
    }
  }
})

export default translate()(withRouter(connect(null, mapDispatchToProps)(SelectionBar)))
