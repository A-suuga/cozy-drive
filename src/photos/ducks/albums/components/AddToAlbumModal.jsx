import styles from '../../../styles/addToAlbum'

import React, { Component } from 'react'
import { translate } from 'cozy-ui/react/I18n'
import Modal, { ModalSection } from 'cozy-ui/react/Modal'
import classNames from 'classnames'

import CreateAlbumForm from './CreateAlbumForm'
import SelectAlbumsForm from './SelectAlbumsForm'
import Loading from '../../../components/Loading'

class AddToAlbumModal extends Component {
  render() {
    const {
      t,
      photos,
      data,
      fetchStatus,
      createAlbum,
      addPhotos,
      onDismiss
    } = this.props

    const isFetchingAlbums =
      fetchStatus === 'pending' || fetchStatus === 'loading'

    return (
      <Modal title={t('Albums.add_photos.title')} secondaryAction={onDismiss}>
        <ModalSection className={styles['coz-modal-section']}>
          <div className={classNames(styles['coz-create-album'])}>
            <CreateAlbumForm
              onSubmitNewAlbum={name =>
                createAlbum(name, photos).then(onDismiss)
              }
            />
          </div>
          {isFetchingAlbums ? (
            <Loading loadingType="albums_fetching" />
          ) : data && data.length > 0 ? (
            <div className={classNames(styles['coz-select-album'])}>
              <SelectAlbumsForm
                albums={{ data, fetchStatus }}
                onSubmitSelectedAlbum={album =>
                  addPhotos(album, photos).then(onDismiss)
                }
              />
            </div>
          ) : null}
        </ModalSection>
      </Modal>
    )
  }
}

export default translate()(AddToAlbumModal)
