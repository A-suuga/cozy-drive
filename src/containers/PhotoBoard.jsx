import styles from '../styles/photoList'

import React, { Component } from 'react'
import { connect } from 'react-redux'
import { translate } from '../lib/I18n'

import { togglePhotoSelection } from '../actions/selection'
import { mustShowSelectionBar } from '../lib/helpers'

import Empty from '../components/Empty'
import Loading from '../components/Loading'
import ErrorComponent from '../components/ErrorComponent'
import SelectionBar from './SelectionBar'
import PhotoList from '../components/PhotoList'
import AddToAlbumModal from '../containers/AddToAlbumModal'

export class PhotoBoard extends Component {
  constructor (props) {
    super(props)
    this.state = {
      isFetching: true,
      photoLists: []
    }

    this.fetchPhotoLists(props)
  }

  componentWillReceiveProps (nextProps, nextState) {
    if (!this.state.isFetching && nextProps.refetch) {
      this.fetchPhotoLists(nextProps)
    }
  }

  fetchPhotoLists (props) {
    props.fetchPhotoLists().then(photoLists => {
      this.setState({
        isFetching: false,
        photoLists: photoLists
      })
    }).catch(photosError => {
      console.error(photosError)
      this.setState({isFetching: false, isError: true})
    })
  }

  render () {
    const {
      showSelection,
      selected,
      showAddToAlbumModal,
      onPhotoToggle,
      photosContext
    } = this.props
    const { isFetching, isWorking, isIndexing } = this.props
    const isGloballyFetching = isFetching || (!isIndexing && this.state.isFetching)
    const { photoLists, isError } = this.state
    const isBusy = isGloballyFetching || isWorking || isIndexing
    if (isError) {
      return <div role='contentinfo'>
        <ErrorComponent errorType={`${photosContext}_photos`} />
      </div>
    }
    return (
      <div
        role='contentinfo'
        className={showSelection ? styles['pho-list-selection'] : ''}
      >
        { isIndexing &&
          <Loading loadingType='photos_indexing' />
        }
        { isGloballyFetching &&
          <Loading loadingType='photos_fetching' />
        }
        { isWorking &&
          <Loading loadingType='photos_upload' />
        }
        { showAddToAlbumModal &&
          <AddToAlbumModal />
        }
        {showSelection && <SelectionBar />}
        {!isBusy && photoLists.map(photoList => {
          return (<PhotoList
            key={photoList.title}
            title={photoList.title}
            photos={photoList.photos}
            selected={selected}
            onPhotoToggle={onPhotoToggle}
          />)
        })}
        {!isBusy && photoLists.length === 0 &&
          <Empty emptyType={`${photosContext}_photos`} />
        }
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => ({
  isFetching: state.ui.isFetching,
  isIndexing: state.ui.isIndexing,
  isWorking: state.ui.isWorking,
  selected: state.ui.selected,
  showSelection: mustShowSelectionBar(state),
  showAddToAlbumModal: state.ui.showAddToAlbumModal
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  onPhotoToggle: (id, selected) => {
    dispatch(togglePhotoSelection(id, selected))
  }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(translate()(PhotoBoard))
