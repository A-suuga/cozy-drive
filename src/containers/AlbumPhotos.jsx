import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'

import { fetchAlbumPhotosStatsById } from '../ducks/albums'

import PhotoBoard from './PhotoBoard'
import Topbar from '../components/Topbar'

export class AlbumPhotos extends Component {
  constructor (props) {
    super(props)
    this.state = {
      photosAreDirty: false
    }
  }

  render () {
    const { router, onFetchAlbumPhotos, album } = this.props
    const albumId = router.params.albumId
    return (
      <div>
        {album.name &&
          <Topbar viewName='albumContent' albumName={album.name} />
        }
        <PhotoBoard
          fetchPhotoLists={() => onFetchAlbumPhotos(albumId)}
          photosContext='album'
        />
        { this.props.children }
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => ({
  album: state.albums.currentAlbum
})

export const mapDispatchToProps = (dispatch, ownProps) => ({
  onFetchAlbumPhotos: (albumId) => {
    return dispatch(fetchAlbumPhotosStatsById(albumId))
      .then(photos => {
        return photos.length ? [{photos}] : []
      })
  }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(AlbumPhotos))
