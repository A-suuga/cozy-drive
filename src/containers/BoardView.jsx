import React, { Component } from 'react'
import { connect } from 'react-redux'

import { isSelectionBarVisible, getSelectedIds, toggleItemSelection } from '../ducks/selection'

import Empty from '../components/Empty'
import Loading from '../components/Loading'
import ErrorComponent from '../components/ErrorComponent'
import SelectionBarWithActions from './SelectionBarWithActions'
import PhotoBoard from '../components/PhotoBoard'
import AddToAlbumModal from '../containers/AddToAlbumModal'

class BoardView extends Component {
  render () {
    const {
      showSelection,
      selected,
      isAddToAlbumModalOpened,
      onPhotoToggle,
      selectionModeActive,
      photosContext
    } = this.props

    const {
      photoLists,
      fetchStatus,
      hasMore,
      onFetchMore
    } = this.props

    return (
      <div role='contentinfo'>
        {isAddToAlbumModalOpened && <AddToAlbumModal />}
        {selectionModeActive && <SelectionBarWithActions />}
        <PhotoBoard
          lists={photoLists}
          selected={selected}
          photosContext={photosContext}
          showSelection={selectionModeActive}
          onPhotoToggle={onPhotoToggle}
          fetchStatus={fetchStatus}
          hasMore={hasMore}
          onFetchMore={onFetchMore}
        />
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => ({
  selected: getSelectedIds(state),
  selectionModeActive: isSelectionBarVisible(state),
  isAddToAlbumModalOpened: state.ui.isAddToAlbumModalOpened
})

const mapDispatchToProps = (dispatch, ownProps) => ({
  onPhotoToggle: (id, selected) => {
    dispatch(toggleItemSelection(id, selected))
  }
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BoardView)
