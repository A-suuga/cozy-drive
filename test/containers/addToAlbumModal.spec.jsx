'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import { mockT } from '../lib/I18n'
import { AddToAlbumModal } from '../../src/containers/AddToAlbumModal'

const mockFetchedPhotos = [
  {
    _id: '33dda00f0eec15bc3b3c59a615001ac8',
    created_at: '0001-01-01T00:00:00Z',
    name: 'MonImage.jpg',
    size: '150000',
    updated_at: '0001-01-01T00:00:00Z'
  }
]

describe('AddToAlbumModal component', () => {
  it('should be rendered correctly', () => {
    const component = shallow(
      <AddToAlbumModal t={mockT} photos={mockFetchedPhotos} onDismiss={jest.fn()} onSubmitNewAlbum={jest.fn()} />
    )
    expect(component.node).toMatchSnapshot()
  })
})
