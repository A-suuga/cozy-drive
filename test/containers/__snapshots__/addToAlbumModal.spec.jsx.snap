exports[`AddToAlbumModal component should be rendered correctly 1`] = `
<Unknown
  secondaryAction={[Function]}
  title="Add to album">
  <div
    className="coz-modal-section coz-create-album">
    <_translate
      onSubmitNewAlbum={[Function]} />
  </div>
</Unknown>
`;
